### PR TITLE
allow multiple authorizers by specifying name and arn

### PR DESCRIPTION
**Pull request by kminkler**
## What did you implement:

Ran into issue specifying multiple authorizers (by arn) because the authorizer names were similar (both ended in the same -suffix)

change Allow for multiple authorizers (with similar names) to be specified by arn by allowing to name each authorizer, e.g

```yaml
authorizer:
   arn: arn:aws:...
   name: authorizer1

authorizer:
   arn: arn:aws:...
   name: authorizer2
```

## How did you implement it:

Check to see if name is specified along with arn.  if so, favor specified name over generated name.

## How can we verify it:

Create two authorizer functions with similar suffixed names, e.g.  auth1-dev  and auth2-dev
create two functions authorized by each authorizer, eg.

```yaml
functions:
  function1:
      handler: handler.function1
      events:
         - http:
             path: /one
             authorizer:
                 arn:  arn:...:auth1-dev
  function2:
      handler: handler.function1
      events:
         - http:
             path: /one
             authorizer:
                 arn:  arn:...:auth2-dev
 ```

The resulting template/api gateway will have only a single authorizer named 'dev' used by both functions, which is not the intended behavior.

Now adding 'name' to the authorizer naming each one uniquely will result in two authorizers being created and used by the functions, as intended:

```yaml
functions:
  function1:
      handler: handler.function1
      events:
         - http:
             path: /one
             authorizer:
                 arn:  arn:...:auth1-dev
                 name: auth1
  function2:
      handler: handler.function1
      events:
         - http:
             path: /one
             authorizer:
                 arn:  arn:...:auth2-dev
                 name: auth2
```

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by pmuens** [comment]
Nice! Thanks for the addition!

Could you please fix the linting error (line is too long) so that the build passes. Thanks!

**Comment by HyperBrain** [comment]
Just created https://github.com/serverless/serverless/issues/3413 which seems the problem.
I mentioned some examples that I experienced there.

@kminkler Your fix should also respect the aliased arns and disambiguate them properly. Additionally all other ways to declare authorizers should lead to unambiguous names too.

File: lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js [hunk 1]
```diff
@@ -202,7 +202,7 @@ module.exports = {
     } else if (typeof authorizer === 'object') {
       if (authorizer.arn) {
         arn = authorizer.arn;
-        name = this.provider.naming.extractAuthorizerNameFromArn(arn);
+        name = typeof authorizer.name == 'string' ? authorizer.name : this.provider.naming.extractAuthorizerNameFromArn(arn);
```
Comment (HyperBrain) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js]: Either use '===' or _.isString(authorizer.name)

Comment (HyperBrain) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js]: There should also be fixes in `extractAuthorizerNameFromArn()` as the malicious name generation happens there in the first place

**Comment by thomasmichaelwallace** [comment]
I was just about to put together a PR to fix this issue myself. Is there anything I can do to help get this through faster?

**Comment by horike37** [comment]
Closing this issue for now since it took for a long time without any activity.
Please ping us if you want to continue working on this!

**Pull request closed without merging**